File: java/src/com/voxgig/sekreto/Json.java
// Minimal JSON support for sekreto.
//
// sekreto adds no third-party dependencies, so it carries just enough JSON
// to read a vault's answer and write the CLI's own line of output. It is
// deliberately not a general-purpose library.
//
// Parses into plain Java values:
//   object  -> LinkedHashMap<String,Object> (insertion ordered)
//   array   -> ArrayList<Object>
//   string  -> String
//   number  -> Double
//   boolean -> Boolean
//   null    -> null

package com.voxgig.sekreto;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class Json {

  private Json() {}

  private static final class Reader {
    private final String text;
    private int at;

    Reader(String text) {
      this.text = text;
    }

    void skip() {
      while (at < text.length() && Character.isWhitespace(text.charAt(at))) {
        at++;
      }
    }

    char peek() {
      return at < text.length() ? text.charAt(at) : '\0';
    }

    boolean done() {
      return at >= text.length();
    }

    Object value() {
      skip();

      if (done()) {
        return null;
      }

      char head = peek();

      if ('{' == head) {
        return object();
      }
      if ('[' == head) {
        return array();
      }
      if ('"' == head) {
        return string();
      }
      if (text.startsWith("true", at)) {
        at += 4;
        return Boolean.TRUE;
      }
      if (text.startsWith("false", at)) {
        at += 5;
        return Boolean.FALSE;
      }
      if (text.startsWith("null", at)) {
        at += 4;
        return null;
      }

      return number();
    }

    Map<String, Object> object() {
      Map<String, Object> out = new LinkedHashMap<>();
      at++; // {
      skip();

      if ('}' == peek()) {
        at++;
        return out;
      }

      while (!done()) {
        skip();
        String key = string();
        skip();
        at++; // :
        out.put(key, value());
        skip();

        if (',' == peek()) {
          at++;
          continue;
        }

        at++; // }
        break;
      }

      return out;
    }

    List<Object> array() {
      List<Object> out = new ArrayList<>();
      at++; // [
      skip();

      if (']' == peek()) {
        at++;
        return out;
      }

      while (!done()) {
        out.add(value());
        skip();

        if (',' == peek()) {
          at++;
          continue;
        }

        at++; // ]
        break;
      }

      return out;
    }

    String string() {
      StringBuilder out = new StringBuilder();
      at++; // opening quote

      while (at < text.length()) {
        char head = text.charAt(at++);

        if ('"' == head) {
          break;
        }

        if ('\\' != head) {
          out.append(head);
          continue;
        }

        char next = text.charAt(at++);
        switch (next) {
          case 'n':
            out.append('\n');
            break;
          case 'r':
            out.append('\r');
            break;
          case 't':
            out.append('\t');
            break;
          case 'b':
            out.append('\b');
            break;
          case 'f':
            out.append('\f');
            break;
          case 'u':
            out.append((char) Integer.parseInt(text.substring(at, at + 4), 16));
            at += 4;
            break;
          default:
            out.append(next);
        }
      }

      return out.toString();
    }

    Double number() {
      int start = at;

      while (at < text.length() && 0 > "]},\n\r\t ".indexOf(text.charAt(at))) {
        at++;
      }

      try {
        return Double.valueOf(text.substring(start, at));
      } catch (NumberFormatException err) {
        return null;
      }
    }
  }

  /** Parse JSON text, answering null for anything unreadable. */
  public static Object parse(String text) {
    if (null == text || text.isEmpty()) {
      return null;
    }

    try {
      return new Reader(text).value();
    } catch (RuntimeException err) {
      return null;
    }
  }

  /** Render a value as compact JSON. */
  public static String stringify(Object value) {
    StringBuilder out = new StringBuilder();
    write(value, out);
    return out.toString();
  }

  private static void write(Object value, StringBuilder out) {
    if (null == value) {
      out.append("null");
      return;
    }

    if (value instanceof String) {
      quote((String) value, out);
      return;
    }

    if (value instanceof Boolean || value instanceof Number) {
      out.append(String.valueOf(value));
      return;
    }

    if (value instanceof Map) {
      out.append('{');
      boolean first = true;
      for (Map.Entry<?, ?> entry : ((Map<?, ?>) value).entrySet()) {
        if (!first) {
          out.append(',');
        }
        first = false;
        quote(String.valueOf(entry.getKey()), out);
        out.append(':');
        write(entry.getValue(), out);
      }
      out.append('}');
      return;
    }

    if (value instanceof List) {
      out.append('[');
      boolean first = true;
      for (Object item : (List<?>) value) {
        if (!first) {
          out.append(',');
        }
        first = false;
        write(item, out);
      }
      out.append(']');
      return;
    }

    quote(String.valueOf(value), out);
  }

  private static void quote(String text, StringBuilder out) {
    out.append('"');

    for (int index = 0; index < text.length(); index++) {
      char head = text.charAt(index);
      switch (head) {
        case '"':
          out.append("\\\"");
          break;
        case '\\':
          out.append("\\\\");
          break;
        case '\n':
          out.append("\\n");
          break;
        case '\r':
          out.append("\\r");
          break;
        case '\t':
          out.append("\\t");
          break;
        default:
          if (0x20 > head) {
            out.append(String.format("\\u%04x", (int) head));
          } else {
            out.append(head);
          }
      }
    }

    out.append('"');
  }
}

File: clojure/src/voxgig/sekreto/json.clj
;; Minimal JSON support for sekreto.
;;
;; sekreto adds no third-party dependencies, so it carries just enough JSON
;; to read a vault's answer and write the CLI's own line of output. It is
;; deliberately not a general-purpose library, and it is not
;; clojure.data.json.
;;
;; Values are plain Clojure data - maps with string keys, vectors, strings,
;; doubles, booleans and nil - with ONE addition. Clojure has nil for JSON
;; null, so "there is no value here" needs a marker of its own: NONE. A
;; store answering `null`, and a store answering text that is not JSON at
;; all, mean different things, and only the second is a store that could not
;; answer coherently. `parse` returns NONE for the second and nil for the
;; first, which is the distinction the callers of fetchjson need.
;;
;; A port of typescript/src/Json.ts, which is canonical.

(ns voxgig.sekreto.json
  (:require [clojure.string :as string]))

;; No value at all: text that is not JSON, or a walk that ran off the end of
;; an object. Never a value a store legitimately sent.
(def NONE ::none)

(defn none?
  "Is this the no-value marker, rather than a value (nil included)?"
  [value]
  (identical? NONE value))

(defn omap
  "An insertion-ordered map, built in one go from key/value pairs.

  `array-map` keeps insertion order, but only until it is grown past eight
  entries by `assoc` - after that it promotes to a hash map, whose iteration
  order is its own. Redaction order and a signed payload's field order both
  depend on the order a map was built in, so ordered maps here are built
  from their pairs rather than assoc'd up one at a time. A repeated key
  keeps its first position and takes its last value, the way an object
  literal does."
  [pairs]
  (loop [keys [] values {} rest (seq pairs)]
    (if (nil? rest)
      (apply array-map (mapcat (fn [key] [key (get values key)]) keys))
      (let [[key value] (first rest)]
        (recur (if (contains? values key) keys (conj keys key))
               (assoc values key value)
               (next rest))))))

(defn numstr
  "Render a number the way every other port does: a whole number has no
  fractional tail, so a JSON `1` read back and printed stays `1`."
  [value]
  (let [number (double value)]
    (cond
      (or (Double/isNaN number) (Double/isInfinite number)) "null"
      (and (== number (Math/rint number)) (> 9007199254740992.0 (Math/abs number)))
      (str (long number))
      :else (str number))))

(defn quotestr
  "Render a string as a JSON string literal, quotes included."
  [^String text]
  (let [out (StringBuilder. "\"")]
    (doseq [ch text]
      (cond
        (= \" ch) (.append out "\\\"")
        (= \\ ch) (.append out "\\\\")
        (= \newline ch) (.append out "\\n")
        (= \return ch) (.append out "\\r")
        (= \tab ch) (.append out "\\t")
        (> 0x20 (int ch)) (.append out (format "\\u%04x" (int ch)))
        :else (.append out ch)))
    (.toString (.append out \"))))

(defn stringify
  "Render a value as compact JSON, keys in the order the map holds them."
  [value]
  (cond
    (none? value) "null"
    (nil? value) "null"
    (boolean? value) (if value "true" "false")
    (string? value) (quotestr value)
    (number? value) (numstr value)
    (vector? value) (str "[" (string/join "," (map stringify value)) "]")
    (map? value) (str "{"
                      (string/join
                       ","
                       (map (fn [[key entry]] (str (quotestr (str key)) ":" (stringify entry)))
                            value))
                      "}")
    :else (quotestr (str value))))

;; ------------------------------------------------------------- the reader
;;
;; Each step takes the text and a position and answers [value position].
;; Malformed input throws, and `parse` - the only entry point - turns that
;; into NONE, so a caller never sees this exception.

(defn- bad [message]
  (throw (ex-info (str "sekreto: json: " message) {:json true})))

(defn- ws? [ch]
  (or (= \space ch) (= \tab ch) (= \newline ch) (= \return ch)))

(defn- skipws [^String text pos]
  (loop [at pos]
    (if (and (> (.length text) at) (ws? (.charAt text at))) (recur (inc at)) at)))

(declare readvalue)

(defn- readword [^String text pos ^String word value]
  (if (and (>= (.length text) (+ pos (count word)))
           (= word (subs text pos (+ pos (count word)))))
    [value (+ pos (count word))]
    (bad (str "bad literal at " pos))))

(defn- readstr [^String text pos]
  (when (or (<= (.length text) pos) (not= \" (.charAt text pos)))
    (bad (str "expected string at " pos)))

  (loop [at (inc pos)
         out (StringBuilder.)]
    (when (<= (.length text) at)
      (bad "unterminated string"))

    (let [ch (.charAt text at)]
      (cond
        (= \" ch) [(.toString out) (inc at)]

        (not= \\ ch) (recur (inc at) (.append out ch))

        (<= (.length text) (inc at)) (bad "unterminated string")

        :else
        (let [escape (.charAt text (inc at))]
          (case escape
            \" (recur (+ at 2) (.append out \"))
            \\ (recur (+ at 2) (.append out \\))
            \/ (recur (+ at 2) (.append out \/))
            \b (recur (+ at 2) (.append out \backspace))
            \f (recur (+ at 2) (.append out \formfeed))
            \n (recur (+ at 2) (.append out \newline))
            \r (recur (+ at 2) (.append out \return))
            \t (recur (+ at 2) (.append out \tab))
            \u (if (< (.length text) (+ at 6))
                 (bad "bad unicode escape")
                 (recur (+ at 6)
                        (.append out (char (Integer/parseInt (subs text (+ at 2) (+ at 6)) 16)))))
            (bad (str "bad escape [" escape "] at " at))))))))

(defn- numchar? [ch]
  (or (Character/isDigit ^char ch)
      (= \. ch) (= \e ch) (= \E ch) (= \- ch) (= \+ ch)))

(defn- readnum [^String text pos]
  (let [end (loop [at pos]
              (if (and (> (.length text) at) (numchar? (.charAt text at))) (recur (inc at)) at))
        span (subs text pos end)]
    (try
      [(Double/parseDouble span) end]
      (catch NumberFormatException _ (bad (str "bad number [" span "] at " pos))))))

(defn- readarr [^String text pos]
  (let [after (skipws text (inc pos))]
    (if (and (> (.length text) after) (= \] (.charAt text after)))
      [[] (inc after)]
      (loop [at after
             out []]
        (let [[value next] (readvalue text (skipws text at))
              done (skipws text next)]
          (when (<= (.length text) done)
            (bad "unterminated array"))
          (case (.charAt text done)
            \, (recur (inc done) (conj out value))
            \] [(conj out value) (inc done)]
            (bad (str "expected ',' or ']' at " done))))))))

(defn- readobj [^String text pos]
  (let [after (skipws text (inc pos))]
    (if (and (> (.length text) after) (= \} (.charAt text after)))
      [(omap []) (inc after)]
      (loop [at after
             pairs []]
        (let [[key next] (readstr text (skipws text at))
              colon (skipws text next)]
          (when (or (<= (.length text) colon) (not= \: (.charAt text colon)))
            (bad (str "expected ':' at " colon)))

          (let [[value after-value] (readvalue text (skipws text (inc colon)))
                done (skipws text after-value)]
            (when (<= (.length text) done)
              (bad "unterminated object"))
            (case (.charAt text done)
              \, (recur (inc done) (conj pairs [key value]))
              \} [(omap (conj pairs [key value])) (inc done)]
              (bad (str "expected ',' or '}' at " done)))))))))

(defn- readvalue [^String text pos]
  (when (<= (.length text) pos)
    (bad "unexpected end"))

  (case (.charAt text pos)
    \{ (readobj text pos)
    \[ (readarr text pos)
    \" (readstr text pos)
    \t (readword text pos "true" true)
    \f (readword text pos "false" false)
    \n (readword text pos "null" nil)
    (readnum text pos)))

(defn parse
  "Parse JSON text. NONE for anything unreadable - which the caller must
  tell apart from a literal `null` body, since only the first means the
  store could not answer coherently."
  [text]
  (if (or (nil? text) (= "" text))
    NONE
    (try
      (let [[value pos] (readvalue text (skipws text 0))]
        (if (> (count text) (skipws text pos)) NONE value))
      (catch Exception _ NONE))))

;; ------------------------------------------------------------- the reads
;;
;; Every read answers nil (or NONE, for `dig`) rather than throwing, so a
;; provider can walk a response body it did not write without checking each
;; step. NONE in, NONE out.

(defn dig
  "Walk nested objects; NONE the moment a step is not there."
  [value & keys]
  (reduce (fn [at key]
            (if (and (map? at) (contains? at key))
              (get at key)
              (reduced NONE)))
          value
          keys))

(defn astext
  "This value as the text a caller would print, or nil when there is no
  value at all. A JSON null is \"no value\": every provider here treats it
  as a miss rather than as the string \"null\"."
  [value]
  (cond
    (none? value) nil
    (nil? value) nil
    (string? value) value
    (boolean? value) (if value "true" "false")
    (number? value) (numstr value)
    :else (stringify value)))

(defn asstr [value] (when (string? value) value))

(defn asnum [value] (when (and (number? value) (not (boolean? value))) value))

(defn asarr [value] (when (vector? value) value))

(defn asobj [value] (when (map? value) value))
